File: packages/js-core/src/queries/feed-query.ts
import { PolycentricClient } from '../polycentric-client';
import { Event, Events, SignedEvent } from '../proto/polycentric';
import { ServerError } from '../utils';
import { Base64 } from 'js-base64';

export interface ResultEventsAndServerErrors {
  events: Events;
  errors: ServerError[];
}

export class FeedQuery {
  private readonly cursors: Map<string, Uint8Array> = new Map();
  private latestEvent?: Event;

  private readonly result: Set<string> = new Set();

  constructor(
    _client: PolycentricClient, // Kept here because we might need access to the client object at some point in the future
    private readonly feedCallback: (
      cursors: Map<string, Uint8Array>,
      latestEvent?: Event,
    ) => Promise<ResultEventsAndServerErrors>,
  ) {}

  public async read(): Promise<ResultEventsAndServerErrors> {
    let result = await this.feedCallback(this.cursors, this.latestEvent);

    let events = result.events.events.filter((signedEvent: SignedEvent) => {
      let event = Base64.fromUint8Array(signedEvent.event);
      return !this.result.has(event);
    });

    for (let signedEvent of events) {
      let event = Base64.fromUint8Array(signedEvent.event);
      this.result.add(event);
    }

    if (events.length > 0) {
      let latestSignedEvent = events[events.length - 1];
      this.latestEvent = Event.fromBinary(latestSignedEvent.event);
    }

    return result;
  }
}

File: packages/kotlin/app/src/main/java/tech/futo/libPolycentric/queries/FeedQuery.kt
package tech.futo.libPolycentric.queries

import android.util.Base64
import polycentric.Event
import polycentric.Events
import polycentric.SignedEvent
import tech.futo.libPolycentric.PolycentricClient

data class ServerError(
    val server: String,
    val error: String,
)

data class ResultEventsAndServerErrors(
    val events: Events,
    val errors: List<ServerError>,
)

class FeedQuery(
    private val client: PolycentricClient,
    private val feedCallback: (cursors: MutableMap<String, ByteArray>, latestEvent: Event?) -> ResultEventsAndServerErrors,
) {
    private val cursors: MutableMap<String, ByteArray> = mutableMapOf()
    private var latestEvent: Event? = null
    private val result: MutableSet<String> = mutableSetOf()

    fun read(): ResultEventsAndServerErrors {
        val callbackResult = feedCallback(cursors, latestEvent)

        val newEvents = callbackResult.events.events.filter { signedEvent ->
            val encoded = Base64.encodeToString(signedEvent.event.toByteArray(), Base64.NO_WRAP)
            !result.contains(encoded)
        }

        for (signedEvent in newEvents) {
            val encoded = Base64.encodeToString(signedEvent.event.toByteArray(), Base64.NO_WRAP)
            result.add(encoded)
        }

        if (newEvents.isNotEmpty()) {
            val latestSignedEvent = newEvents.last()
            latestEvent = Event.ADAPTER.decode(latestSignedEvent.event.toByteArray())
        }

        return callbackResult
    }
}
